File: clients/android/app/src/main/kotlin/com/lextures/android/core/realtime/WebSocketClient.kt
package com.lextures.android.core.realtime

import com.lextures.android.core.config.AppConfiguration
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

/**
 * Lifecycle hooks for per-screen sockets that need connection-state UX (e.g. [BoardSocket]).
 *
 * [httpStatus] is set when the HTTP upgrade failed before the socket opened (e.g. 404).
 * [willReconnect] is false when the client stopped retrying (permanent refusal or [disconnect]).
 */
sealed class WebSocketLifecycleEvent {
    data object Opened : WebSocketLifecycleEvent()
    data class Closed(val httpStatus: Int?, val willReconnect: Boolean) : WebSocketLifecycleEvent()
}

/**
 * Reconnecting JSON WebSocket client for the server's realtime hubs.
 *
 * Mirrors the web client's handshake (`clients/web/src/context/inbox-unread-provider.tsx`):
 * on open, sends `{"authToken":"<jwt>"}` as the first text message (the server reads exactly
 * one auth message before subscribing — see `handleCommWS` in
 * `server/internal/httpserver/communication.go`), then treats every later frame as an event
 * payload. Reconnects after a fixed 2s delay, matching the web app's `scheduleReconnect`.
 *
 * Binary frames are ignored (OkHttp only delivers text to [onMessage] when the String overload
 * is used) — required for board Y.js replay frames (VC.M4).
 */
class WebSocketClient(
    private val path: String,
    private val http: OkHttpClient = OkHttpClient(),
    private val accessTokenProvider: () -> String?,
    private val onMessage: (String) -> Unit,
    private val onLifecycle: ((WebSocketLifecycleEvent) -> Unit)? = null,
    private val stopOnPermanentRefusal: Boolean = false,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var socket: WebSocket? = null
    private var connectedToken: String? = null
    private var isStopped = true
    private var reconnectJob: Job? = null

    /** Connects if not already connected with the current token; safe to call repeatedly. */
    fun connect() {
        isStopped = false
        reconnectJob?.cancel()
        reconnectJob = null
        val token = accessTokenProvider()?.takeIf { it.isNotBlank() } ?: return
        if (socket != null && connectedToken == token) return
        openConnection(token)
    }

    fun disconnect() {
        isStopped = true
        reconnectJob?.cancel()
        reconnectJob = null
        connectedToken = null
        socket?.close(NORMAL_CLOSURE, null)
        socket = null
    }

    private fun openConnection(token: String) {
        socket?.close(NORMAL_CLOSURE, null)
        connectedToken = token

        val url = AppConfiguration.apiUrl(path).toString()
            .replaceFirst("https://", "wss://")
            .replaceFirst("http://", "ws://")
        val request = Request.Builder().url(url).build()

        socket = http.newWebSocket(
            request,
            object : WebSocketListener() {
                override fun onOpen(webSocket: WebSocket, response: Response) {
                    onLifecycle?.invoke(WebSocketLifecycleEvent.Opened)
                    webSocket.send(Json.encodeToString(AuthHandshake(token)))
                }

                override fun onMessage(webSocket: WebSocket, text: String) {
                    onMessage(text)
                }

                override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                    handleDisconnect(webSocket, httpStatus = null)
                }

                override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                    handleDisconnect(webSocket, httpStatus = response?.code)
                }
            },
        )
    }

    private fun handleDisconnect(finished: WebSocket, httpStatus: Int?) {
        if (isStopped || socket !== finished) {
            onLifecycle?.invoke(WebSocketLifecycleEvent.Closed(httpStatus, willReconnect = false))
            return
        }
        socket = null
        connectedToken = null

        val permanent = stopOnPermanentRefusal && isPermanentWsRefusal(httpStatus)
        if (permanent) {
            isStopped = true
            onLifecycle?.invoke(WebSocketLifecycleEvent.Closed(httpStatus, willReconnect = false))
            return
        }

        onLifecycle?.invoke(WebSocketLifecycleEvent.Closed(httpStatus, willReconnect = true))
        reconnectJob = scope.launch {
            delay(RECONNECT_DELAY_MS)
            connect()
        }
    }

    @Serializable
    private data class AuthHandshake(val authToken: String)

    private companion object {
        const val NORMAL_CLOSURE = 1000
        const val RECONNECT_DELAY_MS = 2000L

        fun isPermanentWsRefusal(statusCode: Int?): Boolean =
            statusCode == 401 || statusCode == 403 || statusCode == 404
    }
}

File: clients/ios/Lextures/Core/Realtime/WebSocketClient.swift
import Foundation

/// Shared `URLSession` delegate that signals when a WebSocket handshake completes.
///
/// `URLSessionWebSocketTask` connects asynchronously after `resume()`; sending before
/// `urlSession(_:webSocketTask:didOpenWithProtocol:)` returns POSIX 57 ("Socket is not connected")
/// on device. Android and the web SPA wait for `onOpen` before the auth frame — this does the same.
private final class WebSocketSessionSupport: NSObject, URLSessionWebSocketDelegate, @unchecked Sendable {
    static let shared = WebSocketSessionSupport()

    lazy var session: URLSession = {
        NetworkBootstrap.warmup()
        return URLSession(configuration: .ephemeral, delegate: self, delegateQueue: nil)
    }()

    private let lock = NSLock()
    private var openContinuations: [ObjectIdentifier: CheckedContinuation<Bool, Never>] = [:]
    private var openedTasks: Set<ObjectIdentifier> = []

    /// Suspends until the handshake completes. Returns `true` when `didOpen` fires, `false` when
    /// the task ends before the socket is connected (server down, TLS failure, etc.).
    func waitForOpen(_ task: URLSessionWebSocketTask) async -> Bool {
        await withCheckedContinuation { continuation in
            lock.lock()
            openContinuations[ObjectIdentifier(task)] = continuation
            lock.unlock()
        }
    }

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocol: String?
    ) {
        lock.lock()
        openedTasks.insert(ObjectIdentifier(webSocketTask))
        lock.unlock()
        resumeOpenWaiter(for: webSocketTask, opened: true)
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        guard let webSocketTask = task as? URLSessionWebSocketTask else { return }
        lock.lock()
        let hadOpened = openedTasks.remove(ObjectIdentifier(webSocketTask)) != nil
        lock.unlock()
        guard !hadOpened else { return }
        resumeOpenWaiter(for: webSocketTask, opened: false)
    }

    private func resumeOpenWaiter(for task: URLSessionWebSocketTask, opened: Bool) {
        lock.lock()
        let continuation = openContinuations.removeValue(forKey: ObjectIdentifier(task))
        lock.unlock()
        continuation?.resume(returning: opened)
    }
}

/// Lifecycle hooks for per-screen sockets that need connection-state UX (e.g. BoardSocket).
enum WebSocketLifecycleEvent: Equatable, Sendable {
    case opened
    /// `httpStatus` is set when the HTTP upgrade failed before the socket opened (e.g. 404).
    /// `willReconnect` is false when the client stopped retrying (permanent refusal or disconnect()).
    case closed(httpStatus: Int?, willReconnect: Bool)
}

/// Reconnecting JSON WebSocket client for the server's realtime hubs.
///
/// Mirrors the web client's handshake (`clients/web/src/context/inbox-unread-provider.tsx`):
/// on open, sends `{"authToken":"<jwt>"}` as the first text message (the server reads exactly
/// one auth message before subscribing — see `handleCommWS` in
/// `server/internal/httpserver/communication.go`), then treats every later frame as an event
/// payload. Reconnects after a fixed 2s delay, matching the web app's `scheduleReconnect`.
@MainActor
final class WebSocketClient {
    private let path: String
    private let accessTokenProvider: () -> String?
    private let onMessage: (Data) -> Void
    private let onLifecycle: ((WebSocketLifecycleEvent) -> Void)?
    private let stopOnPermanentRefusal: Bool

    private var task: URLSessionWebSocketTask?
    private var connectedToken: String?
    private var isStopped = true
    private var reconnectTask: Task<Void, Never>?

    init(
        path: String,
        accessTokenProvider: @escaping () -> String?,
        onMessage: @escaping (Data) -> Void,
        onLifecycle: ((WebSocketLifecycleEvent) -> Void)? = nil,
        stopOnPermanentRefusal: Bool = false
    ) {
        self.path = path
        self.accessTokenProvider = accessTokenProvider
        self.onMessage = onMessage
        self.onLifecycle = onLifecycle
        self.stopOnPermanentRefusal = stopOnPermanentRefusal
    }

    /// Connects if not already connected with the current token. Safe to call repeatedly
    /// (e.g. on scenePhase changes or token refresh) — it's a no-op when already connected.
    func connect() {
        isStopped = false
        reconnectTask?.cancel()
        reconnectTask = nil
        guard let token = accessTokenProvider(), !token.isEmpty else { return }
        if task != nil, connectedToken == token { return }
        openConnection(token: token)
    }

    func disconnect() {
        isStopped = true
        reconnectTask?.cancel()
        reconnectTask = nil
        connectedToken = nil
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
    }

    private func openConnection(token: String) {
        task?.cancel(with: .normalClosure, reason: nil)
        connectedToken = token

        let newTask = WebSocketSessionSupport.shared.session.webSocketTask(
            with: AppConfiguration.webSocketURL(path: path)
        )
        task = newTask
        newTask.resume()

        guard let authData = try? JSONSerialization.data(withJSONObject: ["authToken": token]),
              let authString = String(data: authData, encoding: .utf8) else {
            return
        }

        Task { [weak self] in
            guard let self else { return }
            do {
                let opened = await WebSocketSessionSupport.shared.waitForOpen(newTask)
                guard opened, !self.isStopped, self.task === newTask else {
                    let status = (newTask.response as? HTTPURLResponse)?.statusCode
                    await self.handleDisconnect(after: newTask, httpStatus: status)
                    return
                }
                self.onLifecycle?(.opened)
                try await newTask.send(.string(authString))
                try await self.receiveLoop(task: newTask)
                await self.handleDisconnect(after: newTask, httpStatus: nil)
            } catch {
                let status = (newTask.response as? HTTPURLResponse)?.statusCode
                await self.handleDisconnect(after: newTask, httpStatus: status)
            }
        }
    }

    private func receiveLoop(task: URLSessionWebSocketTask) async throws {
        while true {
            let message = try await task.receive()
            switch message {
            case .data(let data):
                onMessage(data)
            case .string(let text):
                if let data = text.data(using: .utf8) {
                    onMessage(data)
                }
            @unknown default:
                break
            }
        }
    }

    private func handleDisconnect(after finishedTask: URLSessionWebSocketTask, httpStatus: Int?) async {
        guard !isStopped, task === finishedTask else {
            onLifecycle?(.closed(httpStatus: httpStatus, willReconnect: false))
            return
        }
        task = nil
        connectedToken = nil

        let permanent = stopOnPermanentRefusal && Self.isPermanentWsRefusal(httpStatus)
        if permanent {
            isStopped = true
            onLifecycle?(.closed(httpStatus: httpStatus, willReconnect: false))
            return
        }

        onLifecycle?(.closed(httpStatus: httpStatus, willReconnect: true))
        reconnectTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(2))
            guard let self, !Task.isCancelled else { return }
            self.connect()
        }
    }

    private static func isPermanentWsRefusal(_ statusCode: Int?) -> Bool {
        guard let statusCode else { return false }
        return statusCode == 401 || statusCode == 403 || statusCode == 404
    }
}
